logstash-core/lib/logstash/agent_plugin_registry.rb

```diff
@@ -0,0 +1,66 @@
+# encoding: utf-8
+require "logstash/agent"
+
+module LogStash module AgentPluginRegistry
+  DEFAULT_AGENT_NAME = :default
+  class DuplicatePluginError < Error; end
+
+  REGISTRY = {}
+
+  # Reset plugin registry to just the default plugin
+  def self.reset!
+    REGISTRY.clear
+    REGISTRY[DEFAULT_AGENT_NAME] = LogStash::Agent
+  end
+
+  reset!
```

Comment:
having a singleton pattern implemented in a module is IMO perfectly acceptable, there is no problem in keeping state, it could even be with using an instance variable in the module. it is a correct pattern.

personally I like using `extend self` instead of defining methods with `self.xxx` but both are acceptable.